protocols/gossipsub/src/behaviour.rs

```diff
@@ -2901,26 +2948,52 @@ where
                 )
             }
         }
+
+        // By default we assume a peer is only a floodsub peer.
+        //
+        // The protocol negotiation occurs once a message is sent/received. Once this happens we
+        // update the type of peer that this is in order to determine which kind of routing should
+        // occur.
+        self.peer_protocols
+            .entry(*peer_id)
+            .or_insert(PeerConnections {
+                kind: PeerKind::Floodsub,
+                connections: vec![*connection_id],
+            })
+            .connections
+            .push(*connection_id);
     }
 
     fn inject_connection_closed(
         &mut self,
-        peer: &PeerId,
-        _: &ConnectionId,
+        peer_id: &PeerId,
+        connection_id: &ConnectionId,
         endpoint: &ConnectedPoint,
     ) {
         // Remove IP from peer scoring system
         if let Some((peer_score, ..)) = &mut self.peer_score {
             if let Some(ip) = get_ip_addr(endpoint.get_remote_address()) {
-                peer_score.remove_ip(peer, &ip);
+                peer_score.remove_ip(peer_id, &ip);
             } else {
                 trace!(
                     "Couldn't extract ip from endpoint of peer {} with endpoint {:?}",
-                    peer,
+                    peer_id,
                     endpoint
                 )
             }
         }
+
+        // Remove the connection from the list
+        // If there are no connections left, inject_disconnected will remove the mapping entirely.
+        if let Some(connections) = self.peer_protocols.get_mut(peer_id) {
+            if let Some(index) = connections
+                .connections
+                .iter()
+                .position(|v| v == connection_id)
+            {
+                connections.connections.remove(index);
+            }
```

Comment:
```suggestion
            let index = connections
                .connections
                .iter()
                .position(|v| v == connection_id)
                .expect("Previously established connection to non-black-listed peer to be present");
            connections.connections.remove(index);
```

As far as I can tell the connection ID should always be present in `connections` right?